codegen/masm/intrinsics/i32.masm

```diff
@@ -0,0 +1,321 @@
+    dup.0 exec.is_signed # [is_b_signed, b, a]
+
+    # is `a` signed?
+    dup.2 exec.is_signed # [is_a_signed, is_b_signed, b, a]
+
+    # compute result
+    movup.3 movup.3 u32wrapping_mul
+
+    # if `b` and `result` have the same sign, which is different than the sign of `a`,
+    # then overflow has occurred. In all other cases, the result is valid
+    movup.3         # [is_b_signed, is_result_signed, result, is_a_signed]
+    dup.1 eq       # [do_b_and_result_have_same_sign, is_result_signed, result, is_a_signed]
+    swap.1 movup.3 # [is_a_signed, is_result_signed, do_b_and_result_have_same_sign, result]
+    neq             # [do_a_and_result_have_different_sign, do_b_and_result_have_same_sign, result]
+    and             # [overflowed, result]
+end
+
+# Multiplies `a` by `b`, asserting that both inputs are valid i32,
+# and asserting that the multiplication does not underflow.
+export.checked_mul # [b, a]
+    exec.overflowing_mul # [overflowed, result]
+    assertz # [result]
+end
+
+# Divides `a` by `b`, asserting that both inputs are valid i32
+export.checked_div # [b, a]
+    u32assert2
+
+    # get positive dividend
+    dup.1 exec.unchecked_neg    # [-a, b, a]
+    dup.2 swap.1                # [-a, a, b, a]
+    movup.3 exec.is_signed      # [is_a_signed, -a, a, b]
+    dup.0 movdn.4 cdrop        # [|a|, b, is_a_signed]
+
+    # get positive divisor
+    dup.1 exec.unchecked_neg    # [-b, |a|, b, is_a_signed]
+    dup.2 swap.1                # [-b, b, |a|, b, is_a_signed]
+    movup.3 exec.is_signed      # [is_b_signed, -b, b, |a|, is_a_signed]
+    dup.0 movdn.4 cdrop        # [|b|, |a|, is_a_signed, is_b_signed]
+
+    # divide
+    u32unchecked_div             # [|a / b|, is_a_signed, is_b_signed]
+
+    # if the signs differ, negate the result
+    movdn.2 neq                 # [signs_differ, |a / b|]
+    dup.1 exec.unchecked_neg    # [-|a / b|, signs_differ, |a / b|]
+    swap.1 cdrop                # [result]
+end
+
+# Given two i32 values in two's complement representation, compare them,
+# returning -1 if `a` < `b`, 0 if equal, and 1 if `a` > `b`.
+#
+export.icmp # [b, a]
+    dup.1 # [a, b, a]
+    dup.1 # [b, a, b, a]
+
+    # get the most-significant bit of `b`
+    push.SIGN_BIT   # [1<<31, b, a, b, a]
+    u32checked_and  # [b_msb, a, b, a]
+
+    # get the most-significant bit of `a`
+    swap.1          # [a, b_msb, b, a]
+    push.SIGN_BIT   # [1<<31, a, b_msb, b, a]
+    u32checked_and  # [a_msb, b_msb, b, a]
+
+    eq.0             # [a_msb == 0, b_msb, b, a]
+    swap.1 eq.0      # [b_msb == 0, a_msb == 0, b, a]
+    swap.1 dup.1 neq # [a_msb != b_msb, b_msb == 0, b, a]
+
+    # if a_msb != b_msb, then a > b (if a_msb == 0), or a < b (if a_msb == 1)
+    if.true # [b_msb == 0, b, a]
+        movdn.2 drop drop         # [b_msb == 0]
+        push.1 push.NEG1          # [-1, 1, b_msb == 0]
+        swap.2                    # [b_msb == 0, 1, -1]
+        cdrop                     # [1 or -1]
```

Comment:
I think the issue is actually just that the `1` and `-1` are swapped. The comments here show the contents of the stack after applying the instructions on that line. Since we know in this branch of the containing `if.true` that `a_msb != b_msb`, we know that `b_msb == 0` is equivalent to `a_msb == 1`.